### PR TITLE
ROCm optimized layernorm for MI100

### PR DESCRIPTION
**Pull request by amathews-amd**
**Description**: Describe your changes.
Optimized block size on MI100 for layerNorm implementation
Optimized global mem reads on MI100 for gradient of layerNorm

**Motivation and Context**
- Why is this change required? What problem does it solve?
https://ontrack.amd.com/browse/MSRCHA-102 : LayerNorm kernel and its grad is slower on MI100 than on V100.
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by amathews-amd** [comment]
@weixingzhang Please review

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu [hunk 1]
```diff
@@ -504,7 +537,11 @@ void HostLayerNormGradient(
   // compute grad_input
   const uint64_t maxGridY = prop.maxGridSize[1];
   const dim3 blocks1(1, std::min<unsigned int>(static_cast<unsigned int>(n1), static_cast<unsigned int>(maxGridY)), 1);
-  const dim3 threads1(warp_size, 4, 1);
+  dim3 threads1(warp_size, 4, 1);
+#ifdef HIP_VERSION
+  // Optimization for ROCm MI100
+  threads1.y = 2;
```
Comment (weixingzhang) [inline orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu:543]: If you only change block size here, how much gain did you see? I am curious whether the change in function cuComputeGradInput is needed or not.

Comment (amathews-amd) [reply orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu:543]: We saw only marginal improvement with just changing block size for cuComputeGradInput.
image

File: onnxruntime/contrib_ops/cuda/layer_norm_impl.cu [hunk 2]
```diff
@@ -364,7 +364,11 @@ void HostApplyLayerNorm(
   const int warp_size = prop.warpSize;
   ORT_ENFORCE(warp_size == GPU_WARP_SIZE);
 
-  const dim3 threads(warp_size, 4, 1);
+  dim3 threads(warp_size, 4, 1);
+#ifdef HIP_VERSION
```
Comment (amathews-amd) [inline onnxruntime/contrib_ops/cuda/layer_norm_impl.cu:368]: HIP_VERSION needs to be  changed to __HIP_PLATFORM_HCC__

**Comment by amathews-amd** [comment]
@jithunnair-amd @pruthvistony @jeffdaily @sunway513 - please review

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu (hunk 1 above)
Comment (amathews-amd) [reply orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu:543]: We saw only marginal improvement with just changing block size.

File: onnxruntime/contrib_ops/cuda/layer_norm_impl.cu (hunk 2 above)
Comment (amathews-amd) [reply onnxruntime/contrib_ops/cuda/layer_norm_impl.cu]: Done

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu (hunk 1 above)
Comment (amathews-amd) [reply orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu:543]: Changing block size is marginal perf improvement ( < 10%), and its not uniform for all cases we tested.

**Review by weixingzhang — APPROVED** [review]

**Pull request merged**